docs/USAGE-GUIDE.md

```diff
@@ -500,6 +500,224 @@ legacy SHA-1 algorithms in CertificateVerify messages if TLS1.2 has been negotia
+
+Servers in particular usually prefer non-blocking mode. In blocking mode, a single connection
+blocks the thread while waiting for more IO. In non-blocking mode, multiple connections
+can make progress by returning control while waiting for more IO using methods like
+[`poll`](https://linux.die.net/man/2/poll) or [`select`](https://linux.die.net/man/2/select).
+
+To use s2n-tls in non-blocking mode, set the underlying file descriptors as non-blocking. 
+For example:
+```c
+int flags = fcntl(fd, F_GETFL, 0);
+if (flags < 0) return S2N_FAILURE;
+if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0) return S2N_FAILURE;
+```
+
+### Errors and Alerts
+
+If the peer sends an alert, the next call to a read IO method will report **S2N_FAILURE** and
+`s2n_error_get_type()` will return **S2N_ERR_T_ALERT**. The specific alert received
+is available by calling `s2n_connection_get_alert()`.
+
+In TLS1.3, all alerts are fatal. s2n-tls also treats all alerts as fatal in earlier
+version of TLS by default. `s2n_config_set_alert_behavior()` can be called to
+force s2n-tls to treat pre-TLS1.3 warning alerts as not fatal, but that behavior
+is not recommended unless required for compatibility. In the past, attacks against
+TLS have involved manipulating the alert level to disguise fatal alerts as warnings.
+
+If s2n-tls encounters a fatal error, the next call to a write IO method will send
+a close_notify alert to the peer. Except for a few exceptions, s2n-tls does not
+send specific alerts in order to avoid leaking information that could be used for
+a sidechannel attack. To ensure that the alert is sent, `s2n_shutdown()` should
+be called after an error.
+
+### Performing the TLS Handshake
+
+Before application data can be sent or received, an application must perform a handshake
+to establish a TLS connection with the peer.
+
+To perform the handshake, call `s2n_negotiate()` until it either returns **S2N_SUCCESS**
+or returns **S2N_FAILURE** without a **S2N_ERR_T_BLOCKED** error.
+
+### Application Data
+
+After the TLS handshake, an application can send and receive encrypted data.
+
+Although most s2n-tls APIs are not thread-safe, `s2n_send()` and `s2n_recv()`
+may be called simultaneously from two different threads. This means that an
+application may have one thread calling `s2n_send()` and one thread calling `s2n_recv()`,
+but NOT multiple threads calling `s2n_recv()` or multiple threads calling `s2n_send()`. 
+
+Even if an application only intends to send data or only intends to receive data,
+it should implement both send and receive in order to handle alerts and post-handshake
+TLS control messages like session tickets.
+
+#### Sending Application Data
+
+`s2n_send()` and its variants encrypt and send application data to the peer.
+The sending methods return the number of bytes written and may indicate a partial
+write. Partial writes are possible not just for non-blocking I/O, but also for
+connections aborted while active.
```

Comment:
What about writes that exceed the record size? Do we impose any limits smaller than that either?

Reply:
No. We can send multiple records for a write.